pkg/controller/infra/infra_controller.go

```diff
@@ -234,27 +526,154 @@ func (c *InfraController) syncCluster(key string) error {
 	cluster, err := c.clustersLister.Clusters(ns).Get(name)
 	if errors.IsNotFound(err) {
 		cLog.Warnln("cluster not found")
+		c.expectations.DeleteExpectations(key)
 		return nil
 	}
 	if err != nil {
 		return err
 	}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
+	if !c.expectations.SatisfiedExpectations(key) {
+		// expectations have not been met, come back later
+		cLog.Debugln("cluster has not satisfied expectations yet")
+		return nil
+	}
+
+	// Find provisioning job
+	var provisioningJob *v1batch.Job
+	jobs, err := c.jobsLister.Jobs(cluster.Namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, job := range jobs {
```

Comment:
We need to do something about outstanding jobs after the cluster spec changes. I think that we don't want to have two jobs running simultaneously for a cluster, so we need to wait for outstanding jobs to finish before starting new jobs. We need to terminate old jobs so that they don't attempt to create more pods in case the old cluster spec is such that the job will never complete successfully.